In [85]:
import pandas as pd
import numpy as np

## Reading files

In [86]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [87]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [88]:
wh = pd.concat([wh2018,wh2019])

In [89]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [61]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [90]:
# Lista de valores nulos
for c_ in wh.columns:
    print(np.sum(wh[c_].isna()))

0
0
19
12
16
26
23
21
13
0


In [100]:
#Promedio de los valores nulos
prom_nulos = wh.isnull().mean()*100

In [101]:
prom_nulos

Overall rank                    0.000000
Country or region               0.000000
Score                           6.089744
GDP per capita                  3.846154
Social support                  5.128205
Healthy life expectancy         8.333333
Freedom to make life choices    7.371795
Generosity                      6.730769
Perceptions of corruption       4.166667
year                            0.000000
dtype: float64

In [111]:
dataNulos={
    "year":["year"]
    ,"columns":["columns"]
    ,"prom_nulos":["prom_nulos"]
 }

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [10]:
wh.set_index(["year"], inplace=True)

In [103]:
wh.groupby(["year"]).describe()

Overall rank                                                    Score  \
            count  mean        std  min    25%   50%     75%    max  count   
year                                                                         
2018        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  151.0   
2019        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  142.0   

                ... Generosity        Perceptions of corruption            \
          mean  ...        75%    max                     count      mean   
year            ...                                                         
2018  5.354060  ...    0.23400  0.598                     149.0  0.112859   
2019  5.434028  ...    0.24825  0.566                     150.0  0.109053   

                                                     
           std  min      25%    50%      75%    max  
year                                                 
2018  0.098108  0.0  0.05000  0.082  0.13800  0.457  
2019  0.093279  0.0  0.04775  0.085  0.14025  0.453  

[2 rows x 64 columns]

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [ ]:
# Remplazamos los valores nulos con la media

In [24]:
wh = wh.fillna(wh.mean())

C:\Users\ROVINR~1\AppData\Local\Temp/ipykernel_6448/259370468.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  wh = wh.fillna(wh.mean())


In [77]:
#Comprobamos que no hay valores nulos
for c_ in wh.columns:
    print(np.sum(wh[c_].isna()))

0
0
19
12
16
26
23
21
13
0


4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro

In [40]:
#Ordenamos de forma ascendente por Overall rank y Country or region	

wh.sort_values(["Overall rank","Country or region"] , ascending = True )

Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
0               1                   Finland  7.769           1.340   
1               2                   Denmark  7.600           1.383   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
..            ...                       ...    ...             ...   
153           154               South Sudan  3.254           0.337   
154           155  Central African Republic  3.083           0.024   
154           155  Central African Republic  3.083           0.026   
155           156                   Burundi  2.905           0.091   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             1.592                    0.874                         0.681   
0             1.587                      NaN                         0.596   
1             1.573                    0.996                         0.592   
1             1.582                    0.861                         0.686   
2             1.590                    0.868                         0.683   
..              ...                      ...                           ...   
153             NaN                    0.177                         0.112   
154           0.000                    0.010                         0.305   
154           0.000                    0.105                         0.225   
155           0.627                    0.145                         0.065   
155           0.575                    0.295                         0.010   

     Generosity  Perceptions of corruption  year  rank  
0         0.202                      0.393  2018   1.0  
0         0.153                      0.393  2019   2.0  
1         0.252                      0.410  2019   1.0  
1         0.286                      0.340  2018   2.0  
2         0.284                      0.408  2018   1.0  
..          ...                        ...   ...   ...  
153       0.224                      0.106  2018   2.0  
154       0.218                      0.038  2018   1.0  
154       0.235                      0.035  2019   2.0  
155       0.149                      0.076  2018   1.0  
155       0.202                      0.091  2019   2.0  

[312 rows x 11 columns]

In [36]:
wh["rank"] = wh.groupby(["Overall rank"])["Country or region"].rank( method = 'first' , ascending = True )

In [37]:
wh["rank"]

0      1.0
1      2.0
2      1.0
3      1.0
4      2.0
      ... 
151    1.0
152    2.0
153    1.0
154    2.0
155    2.0
Name: rank, Length: 312, dtype: float64

In [43]:
primer_valor = wh.loc[ wh['rank'] == 1 , : ]

In [44]:
primer_valor

Overall rank Country or region  Score  GDP per capita  Social support  \
0               1           Finland  7.632           1.305           1.592   
2               3           Denmark  7.555           1.351           1.590   
3               4           Iceland  7.495           1.343             NaN   
5               6       Netherlands    NaN           1.361           1.488   
6               7            Canada  7.328           1.330           1.532   
..            ...               ...    ...             ...             ...   
146           147             Haiti  3.597           0.323           0.688   
147           148          Botswana  3.488           1.041           1.145   
149           150            Malawi  3.410           0.191           0.560   
151           152            Rwanda  3.334           0.359           0.711   
153           154       Afghanistan  3.203           0.350           0.517   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
0                      0.874                         0.681       0.202   
2                      0.868                         0.683       0.284   
3                      0.914                         0.677         NaN   
5                      0.878                         0.638       0.333   
6                      0.896                         0.653       0.321   
..                       ...                           ...         ...   
146                    0.449                         0.026       0.419   
147                    0.538                         0.455       0.025   
149                    0.495                         0.443       0.218   
151                    0.614                         0.555       0.217   
153                    0.361                         0.000       0.158   

     Perceptions of corruption  year  rank  
0                        0.393  2018   1.0  
2                        0.408  2018   1.0  
3                        0.138  2018   1.0  
5                        0.295  2018   1.0  
6                        0.291  2018   1.0  
..                         ...   ...   ...  
146                      0.110  2019   1.0  
147                      0.100  2019   1.0  
149                      0.089  2019   1.0  
151                      0.411  2019   1.0  
153                      0.025  2019   1.0  

[156 rows x 11 columns]

In [53]:
Ultimo_valor = wh.loc[ wh['rank'] == 2 , : ]

In [54]:
Ultimo_valor

Overall rank         Country or region  Score  GDP per capita  \
1               2                    Norway  7.594           1.456   
4               5               Switzerland  7.487           1.420   
8               9                    Sweden  7.314           1.355   
10             11            United Kingdom  7.190           1.244   
16             17                Luxembourg  6.910           1.576   
..            ...                       ...    ...             ...   
148           149                     Syria  3.462           0.619   
150           151                     Yemen    NaN           0.287   
152           153                  Tanzania  3.231           0.476   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
1             1.582                    0.861                         0.686   
4             1.549                    0.927                         0.660   
8             1.501                    0.913                         0.659   
10            1.433                    0.888                         0.464   
16            1.520                    0.896                         0.632   
..              ...                      ...                           ...   
148           0.378                    0.440                         0.013   
150           1.163                    0.463                         0.143   
152           0.885                    0.499                         0.417   
154           0.000                    0.105                         0.225   
155           0.575                    0.295                         0.010   

     Generosity  Perceptions of corruption  year  rank  
1         0.286                      0.340  2018   2.0  
4         0.256                      0.357  2018   2.0  
8         0.285                      0.383  2018   2.0  
10        0.262                      0.082  2018   2.0  
16          NaN                      0.321  2018   2.0  
..          ...                        ...   ...   ...  
148       0.331                      0.141  2019   2.0  
150       0.108                      0.077  2019   2.0  
152       0.276                      0.147  2019   2.0  
154       0.235                      0.035  2019   2.0  
155       0.202                      0.091  2019   2.0  

[156 rows x 11 columns]